# Databricks Overview 

Databricks provides a Unified Analytics Platform to help business transform their data strategy. We do this by providing a collaborative, scalable, and easy-to-use environment built around Databricks Runtime and Apache Spark.

Databricks provides several key differentiators that will be shown in this demo, including Data engineering capabilities, Databricks Delta,  Databricks Runtime.

In this demo, we'll cover the basics of Databricks using the following architecture:
<div style="text-align: center; line-height: 0; padding-top: 0px;"><br><img src="https://i.imgur.com/5n2ZX8o.png" height="00" width="700"></div>

We'll break this down into 4 sections:
- Platform Overview
- Loading and viewing data
- Manipulating and enriching data
- Streaming data
- Databricks Delta

In [2]:
%run ./Delta-Setup

In [3]:
%fs ls /databricks/ganeshrj/stocks

path name size dbfs:/databricks/ganeshrj/stocks/AMZN.csv AMZN.csv 98657 dbfs:/databricks/ganeshrj/stocks/APPL.csv APPL.csv 102568 dbfs:/databricks/ganeshrj/stocks/GOOG.csv GOOG.csv 103177 dbfs:/databricks/ganeshrj/stocks/NFLX.csv NFLX.csv 1524

In [4]:
 stocksDF = spark.read.jdbc(url=jdbcUrl, table="metadata", properties=connectionProps)
 display (stocksDF)

sticker name APPL Apple Computers Limited GOOG Google Inc NFLX Netflix Inc AMZN Amazon Inc

Run the following code cell to set up paths and remove any leftover data.

In [6]:
genericDataPath = userhome + "/generic/stock-data/"
deltaDataPath = userhome + "/delta/stock-data/"
backfillDataPath = userhome + "/delta/backfill-data/"
deltaMiniDataPath = userhome + "/delta/mini-delta"

dbutils.fs.rm(genericDataPath, True)
dbutils.fs.rm(deltaDataPath, True)
dbutils.fs.rm(backfillDataPath, True)
dbutils.fs.rm(deltaMiniDataPath, True)

Out[ 7 ]: True

###  READ csv data for each stock then WRITE to Parquet / Delta

Read the data into a DataFrame. Since this is a CSV file, let Spark infer the schema from the first row by setting
* `inferSchema` to `true`
* `header` to `true`

Drop the `Adj Close` column, as it's not needed.

## Part 2: Loading and viewing data

Now that we've seen the environment, we can load some data. In this case, we'll load user data from a MySQL database, and also some reference data from a CSV file on S3.
<div style="text-align: center; line-height: 0; padding-top: 9px;"><img src="https://i.imgur.com/r4hY0rS.png" height="00" width="700"></div>

On Databricks, it's simple to unify many different sources across traditional silos into a single environment; it's just a few lines of code.

In [9]:
for stickers in ['APPL', 'GOOG', 'AMZN', 'NFLX']:
  path = "dbfs:/databricks/ganeshrj/stocks/" + stickers + ".csv"
  print (path)  
  rawDataDF = (spark.read 
  .option("inferSchema", "true") 
  .option("header", "true")
  .csv(path)
  .drop("Adj Close"))
  
  rawDataDF = rawDataDF.withColumn("month",(month(rawDataDF.date))).withColumn("Year",(year(rawDataDF.date))).withColumn("Stickers", lit(stickers))
  
  ## Join with StocksDF to get the company name
  joinedDF=rawDataDF.join(broadcast(stocksDF), rawDataDF.Stickers == stocksDF.sticker).drop("Stickers")
  
  ### Write as Parquet 
  joinedDF.write.format("parquet").partitionBy("Year").mode("append").save(genericDataPath)
  
  ### Write as Delta
  joinedDF.write.format("delta").partitionBy("Year").mode("append").save(deltaDataPath)
  
  display (joinedDF)
  
   

--------------------------------------------------------------------------- 
 NameError Traceback (most recent call last)
 <command-2912287> in <module> () 
 8 .drop("Adj Close"))
 9 
 ---> 10 rawDataDF = rawDataDF . withColumn ( "month" , ( month ( rawDataDF . date ) ) ) . withColumn ( "Year" , ( year ( rawDataDF . date ) ) ) . withColumn ( "Stickers" , lit ( stickers ) ) 
 11 
 12 ## Join with StocksDF to get the company name 

 NameError : name 'month' is not defined

In [10]:
rawDataDF.printSchema()

root
-- date: timestamp (nullable = true)
-- open: double (nullable = true)
-- high: double (nullable = true)
-- low: double (nullable = true)
-- close: double (nullable = true)
-- volume: integer (nullable = true)

-sandbox
### CREATE Using Non-Delta Pipeline

Create a table called `generic_stocks` using `parquet` out of the above data.

<img alt="Caution" title="Caution" style="vertical-align: text-bottom; position: relative; height:1.3em; top:0.0em" src="https://files.training.databricks.com/static/images/icon-warning.svg"/> Notice how you MUST specify a schema and partitioning info!

In [12]:
spark.sql("""
    use ngcdemo
  """)
spark.sql("""
    DROP TABLE IF EXISTS generic_stocks
  """)
spark.sql("""
    CREATE TABLE generic_stocks (
      Date TIMESTAMP,
      Open DOUBLE,
      High DOUBLE,
      Low DOUBLE,
      Close DOUBLE,
      Volume INTEGER,
      Month INTEGER,
      Year INTEGER, 
      Sticker STRING,
      name STRING)
    USING parquet 
    OPTIONS (path = '{}' )
    PARTITIONED BY (Year)
  """.format(genericDataPath))
None

Perform a simple `count` query to verify the number of records.

In [14]:
%sql
SELECT count(*) FROM generic_stocks

count(1) 0

-sandbox

<img alt="Caution" title="Caution" style="vertical-align: text-bottom; position: relative; height:1.3em; top:0.0em" src="https://files.training.databricks.com/static/images/icon-warning.svg"/> Wait, no results? 

What is going on here is a problem that stems from its Apache Hive origins.
It's the concept of <br>
<b>schema on read</b> where data is applied to a plan or schema as it is pulled out of a stored location, rather than as it goes into a stored location.

This means that as soon as you put data into a data lake, the schema is unknown <i>until</i> you perform a read operation.

To remedy, you repair the table using `MSCK REPAIR TABLE`.

<img alt="Side Note" title="Side Note" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.05em; transform:rotate(15deg)" src="https://files.training.databricks.com/static/images/icon-note.webp"/> Only after table repair is our count of customer data correct.

In [16]:
%sql
MSCK REPAIR TABLE generic_stocks;

SELECT  Sticker , High, Low  FROM generic_stocks where date(Date) = '2019-03-25' 


Sticker High Low GOOG 1206.4 1187.31 AMZN 1782.68 1747.5 NFLX 367.04 357.44 APPL 191.976 186.6

In [17]:
%sql 

select * from generic_stocks;

Date Open High Low Close Volume Month Sticker name Year 2013-12-31T00:00:00.000+0000 79.1671 80.1829 79.1429 80.1457 55771100 12 APPL Apple Computers Limited 2013 2013-12-30T00:00:00.000+0000 79.6371 80.0129 78.9029 79.2171 63407400 12 APPL Apple Computers Limited 2013 2013-12-27T00:00:00.000+0000 80.5457 80.63 79.9286 80.0129 56471100 12 APPL Apple Computers Limited 2013 2013-12-26T00:00:00.000+0000 81.1571 81.3571 80.4829 80.5571 51002000 12 APPL Apple Computers Limited 2013 2013-12-24T00:00:00.000+0000 81.4129 81.6971 80.8614 81.0957 41888700 12 APPL Apple Computers Limited 2013 2013-12-23T00:00:00.000+0000 81.1429 81.5314 80.3943 81.4414 125326600 12 APPL Apple Computers Limited 2013 2013-12-20T00:00:00.000+0000 77.9186 78.8014 77.8314 78.4314 109103400 12 APPL Apple Computers Limited 2013 2013-12-19T00:00:00.000+0000 78.5 78.5714 77.6757 77.78 80077200 12 APPL Apple Computers Limited 2013 2013-12-18T00:00:00.000+0000 78.5286 78.7786 76.9714 78.6814 141465800 12 APPL Apple Computers Limited 2013 2013-12-17T00:00:00.000+0000 79.4014 79.92 79.0543 79.2843 57475600 12 APPL Apple Computers Limited 2013 2013-12-16T00:00:00.000+0000 79.2886 80.3771 79.2871 79.6429 70648200 12 APPL Apple Computers Limited 2013 2013-12-13T00:00:00.000+0000 80.4071 80.4114 79.0957 79.2043 83205500 12 APPL Apple Computers Limited 2013 2013-12-12T00:00:00.000+0000 80.3057 80.7629 80.0043 80.0771 65572500 12 APPL Apple Computers Limited 2013 2013-12-11T00:00:00.000+0000 81.0 81.5671 79.9557 80.1943 89929700 12 APPL Apple Computers Limited 2013 2013-12-10T00:00:00.000+0000 80.5114 81.1257 80.1714 80.7929 69567400 12 APPL Apple Computers Limited 2013 2013-12-09T00:00:00.000+0000 80.1286 81.3686 80.1286 80.9186 80123400 12 APPL Apple Computers Limited 2013 2013-12-06T00:00:00.000+0000 80.8271 80.9643 79.9386 80.0029 86088100 12 APPL Apple Computers Limited 2013 2013-12-05T00:00:00.000+0000 81.8071 82.1629 80.9157 81.1286 111895000 12 APPL Apple Computers Limited 2013 2013-12-04T00:00:00.000+0000 80.7857 81.3129 80.1171 80.7143 94452400 12 APPL Apple Computers Limited 2013 2013-12-03T00:00:00.000+0000 79.7571 80.9114 79.6686 80.9029 112742000 12 APPL Apple Computers Limited 2013 2013-12-02T00:00:00.000+0000 79.7143 80.6186 78.6886 78.7471 118136200 12 APPL Apple Computers Limited 2013 2013-11-29T00:00:00.000+0000 78.4971 79.7614 78.2586 79.4386 79531900 11 APPL Apple Computers Limited 2013 2013-11-27T00:00:00.000+0000 76.6157 78.0 76.2 77.9943 90862100 11 APPL Apple Computers Limited 2013 2013-11-26T00:00:00.000+0000 74.8743 76.5914 74.8571 76.2 100345700 11 APPL Apple Computers Limited 2013 2013-11-25T00:00:00.000+0000 74.4314 75.1243 74.4286 74.82 57327900 11 APPL Apple Computers Limited 2013 2013-11-22T00:00:00.000+0000 74.2171 74.5943 74.0757 74.2571 55931400 11 APPL Apple Computers Limited 2013 2013-11-21T00:00:00.000+0000 73.9429 74.4586 73.3814 74.4486 65506700 11 APPL Apple Computers Limited 2013 2013-11-20T00:00:00.000+0000 74.1757 74.3457 73.4757 73.5714 48479200 11 APPL Apple Computers Limited 2013 2013-11-19T00:00:00.000+0000 74.1471 74.7686 73.9957 74.2214 52234700 11 APPL Apple Computers Limited 2013 2013-11-18T00:00:00.000+0000 74.9986 75.3129 74.0286 74.09 61236000 11 APPL Apple Computers Limited 2013 2013-11-15T00:00:00.000+0000 75.2257 75.5843 74.9271 74.9986 79480100 11 APPL Apple Computers Limited 2013 2013-11-14T00:00:00.000+0000 74.6871 75.6114 74.5529 75.4514 70604800 11 APPL Apple Computers Limited 2013 2013-11-13T00:00:00.000+0000 74.0 74.6071 73.8514 74.3757 49305200 11 APPL Apple Computers Limited 2013 2013-11-12T00:00:00.000+0000 73.9529 74.8457 73.8571 74.2871 51069200 11 APPL Apple Computers Limited 2013 2013-11-11T00:00:00.000+0000 74.2843 74.5243 73.4871 74.15 56863100 11 APPL Apple Computers Limited 2013 2013-11-08T00:00:00.000+0000 73.5114 74.4471 73.2271 74.3657 69829200 11 APPL Apple Computers Limited 2013 2013-11-07T00:00:00.000+0000 74.2257 74.7414 73.1971 73.2129 65655100 11 APPL Apple Computers Limited 2013 2013-11-

In [18]:
%sql 

select * from stocks_delta;

date open high low close volume month Year sticker name 2014-12-31T00:00:00.000+0000 528.345 529.688 522.925 523.521 1372000 12 2014 GOOG Google Inc 2014-12-30T00:00:00.000+0000 525.202 528.245 524.247 527.519 878600 12 2014 GOOG Google Inc 2014-12-29T00:00:00.000+0000 529.28 532.552 527.113 527.43 2284800 12 2014 GOOG Google Inc 2014-12-26T00:00:00.000+0000 525.878 531.328 524.426 531.11 1043400 12 2014 GOOG Google Inc 2014-12-24T00:00:00.000+0000 527.609 528.851 524.138 525.878 707800 12 2014 GOOG Google Inc 2014-12-23T00:00:00.000+0000 524.118 531.637 523.412 527.688 2203600 12 2014 GOOG Google Inc 2014-12-22T00:00:00.000+0000 513.258 523.581 513.258 522.0 2731200 12 2014 GOOG Google Inc 2014-12-19T00:00:00.000+0000 508.713 514.889 504.139 513.526 3700300 12 2014 GOOG Google Inc 2014-12-18T00:00:00.000+0000 510.145 511.06 501.94 508.305 2934700 12 2014 GOOG Google Inc 2014-12-17T00:00:00.000+0000 494.282 504.227 494.093 502.129 2891000 12 2014 GOOG Google Inc 2014-12-16T00:00:00.000+0000 508.763 510.244 486.326 492.681 3975100 12 2014 GOOG Google Inc 2014-12-15T00:00:00.000+0000 519.881 520.239 510.463 510.99 2821100 12 2014 GOOG Google Inc 2014-12-12T00:00:00.000+0000 520.647 525.61 515.824 515.824 2000000 12 2014 GOOG Google Inc 2014-12-11T00:00:00.000+0000 524.914 531.0 524.218 525.451 1615200 12 2014 GOOG Google Inc 2014-12-10T00:00:00.000+0000 530.165 533.397 522.686 523.183 1716900 12 2014 GOOG Google Inc 2014-12-09T00:00:00.000+0000 519.285 531.269 517.654 530.453 1876400 12 2014 GOOG Google Inc 2014-12-08T00:00:00.000+0000 524.247 528.096 520.926 524.098 2335700 12 2014 GOOG Google Inc 2014-12-05T00:00:00.000+0000 528.096 529.976 521.413 522.388 2572600 12 2014 GOOG Google Inc 2014-12-04T00:00:00.000+0000 528.255 534.402 525.699 534.372 1395900 12 2014 GOOG Google Inc 2014-12-03T00:00:00.000+0000 528.534 533.065 526.366 528.414 1281400 12 2014 GOOG Google Inc 2014-12-02T00:00:00.000+0000 530.593 532.572 526.903 530.831 1530800 12 2014 GOOG Google Inc 2014-12-01T00:00:00.000+0000 535.953 538.449 528.952 530.881 2121100 12 2014 GOOG Google Inc 2014-11-28T00:00:00.000+0000 537.664 539.036 533.666 538.867 1151500 11 2014 GOOG Google Inc 2014-11-26T00:00:00.000+0000 537.922 538.589 534.105 537.415 1527100 11 2014 GOOG Google Inc 2014-11-25T00:00:00.000+0000 536.052 541.005 535.657 538.121 1794800 11 2014 GOOG Google Inc 2014-11-24T00:00:00.000+0000 534.71 539.732 532.691 536.321 1711000 11 2014 GOOG Google Inc 2014-11-21T00:00:00.000+0000 538.648 539.175 533.626 534.561 2230300 11 2014 GOOG Google Inc 2014-11-20T00:00:00.000+0000 528.345 532.184 528.176 531.905 1567600 11 2014 GOOG Google Inc 2014-11-19T00:00:00.000+0000 532.074 535.297 527.181 534.053 1396000 11 2014 GOOG Google Inc 2014-11-18T00:00:00.000+0000 534.561 538.976 531.249 532.104 1968000 11 2014 GOOG Google Inc 2014-11-17T00:00:00.000+0000 540.607 540.816 531.141 533.576 1730700 11 2014 GOOG Google Inc 2014-11-14T00:00:00.000+0000 543.69 543.69 539.185 541.423 1293000 11 2014 GOOG Google Inc 2014-11-13T00:00:00.000+0000 546.793 546.793 540.508 542.398 1343000 11 2014 GOOG Google Inc 2014-11-12T00:00:00.000+0000 547.38 547.45 542.189 544.317 1132700 11 2014 GOOG Google Inc 2014-11-11T00:00:00.000+0000 545.491 548.922 543.313 547.281 968200 11 2014 GOOG Google Inc 2014-11-10T00:00:00.000+0000 538.499 546.585 538.061 544.496 1137700 11 2014 GOOG Google Inc 2014-11-07T00:00:00.000+0000 543.223 543.223 535.724 538.052 1638200 11 2014 GOOG Google Inc 2014-11-06T00:00:00.000+0000 542.517 543.894 538.012 539.076 1336900 11 2014 GOOG Google Inc 2014-11-05T00:00:00.000+0000 553.755 553.755 541.075 542.935 2037800 11 2014 GOOG Google Inc 2014-11-04T00:00:00.000+0000 549.976 552.462 546.296 551.08 1247600 11 2014 GOOG Google Inc 2014-11-03T00:00:00.000+0000 552.462 554.849 550.205 552.184 1386000 11 2014 GOOG Google Inc 2014-10-31T00:00:00.000+0000 556.291 556.51 551.716 556.023 2040600 10 2014 GOOG Google Inc 2014-10-30T00:00:00.000+0000 545.948 549.777 540.538 547.30

## Part 3: Manipulating and Enriching Data

We've loaded some data into Databricks, but now what? Databricks makes it easy to clean, manipulate, enrich, and then store data for downstream use. 

We'll blend our two data sources together, then view the results. We'll also use __Delta__ to persist the data into a permanent, performant, and easily accessible format.

<div style="text-align: center; line-height: 0; padding-top: 9px;"><img src="https://i.imgur.com/N2ASM7K.png" height="00" width="700"></div>

In [20]:
from pyspark.sql.types import * 


sentimentsDF = spark.read.option("inferSchema", "true").option("header", "true").csv("hdfs://hdp-c1.ganeshrj.com:8020/user/root/APPL_Sentiments.csv")


redinedDF=sentimentsDF.withColumnRenamed("published at", "publishTime")

redinedDF.write.format("delta").saveAsTable("ngcdemo.appl_sentiments")


In [21]:
%sql describe ngcdemo.appl_sentiments 

col_name data_type comment title string null ticker string null author string null description string null source string null url string null publishTime string null sentiment double null

In [22]:
%sql 
select count(title) as cnt, date(publishTime) as pdate, avg(sentiment) as senti from ngcdemo.appl_sentiments group by date(publishTime) order by count(title) desc;

cnt pdate senti 102 2019-03-25 0.35960784313725486 66 2019-04-23 0.01803030303030303 66 2019-04-09 0.2872727272727273 60 2019-03-26 0.1465 57 2018-11-28 0.2545614035087719 57 2019-04-25 0.21807017543859641 56 2019-04-26 0.3037499999999999 55 2019-04-27 0.3045454545454546 54 2019-04-10 0.16870370370370372 53 2019-04-24 0.28150943396226413 53 2019-04-22 0.43603773584905653 51 2019-01-30 0.24058823529411763 50 2018-11-16 0.2144 48 2018-11-22 0.153125 46 2019-04-28 0.1541304347826087 44 2018-11-29 0.2413636363636364 44 2019-04-20 0.31522727272727274 42 2018-11-27 0.27999999999999997 40 2018-12-04 0.28474999999999995 35 2018-11-20 0.011142857142857142 31 2019-04-19 0.2735483870967742 29 2018-12-05 0.22586206896551725 28 2018-11-26 0.12 28 2018-11-30 0.14892857142857138 27 2018-11-19 0.15111111111111108 26 2019-01-31 0.08615384615384616 26 2018-12-03 0.2880769230769231 25 2019-04-21 0.26280000000000003 25 2019-01-29 -0.006400000000000001 24 2018-11-17 0.22250000000000003 23 2018-12-07 0.2260869565217391 22 2018-12-01 0.3281818181818182 22 2018-12-08 0.2254545454545455 22 2019-01-09 0.030909090909090907 22 2019-01-03 -0.14727272727272725 20 2019-02-06 0.07649999999999998 18 2018-11-21 0.33555555555555555 17 2019-04-17 0.13117647058823528 16 2019-02-01 -0.07249999999999998 16 2018-12-10 -0.18000000000000002 16 2019-03-20 0.22625 15 2019-03-07 0.24666666666666665 15 2018-11-24 0.3446666666666667 15 2018-12-06 0.2586666666666667 15 2019-03-18 0.34600000000000003 15 2019-03-19 0.28066666666666673 14 2019-04-18 0.42571428571428566 14 2019-03-28 0.35142857142857137 14 2019-03-21 0.4192857142857142 12 2019-03-13 -0.2508333333333333 12 2019-01-04 -0.2541666666666666 12 2019-04-08 0.35500000000000004 11 2019-04-01 0.5509090909090908 11 2019-04-16 0.4263636363636365 11 2019-01-24 0.07 11 2019-03-06 0.1990909090909091 11 2019-03-27 0.36909090909090914 10 2019-01-08 0.329 10 2019-03-12 0.17400000000000002 10 2018-12-11 -0.015000000000000013 10 2019-03-29 0.225 9 2019-01-18 0.008888888888888894 9 2019-02-08 0.4166666666666667 9 2019-01-28 0.0988888888888889 9 2019-01-05 -0.19777777777777775 9 2019-03-14 0.5588888888888888 9 2019-02-13 0.2088888888888889 9 2019-03-22 0.21777777777777774 9 2019-02-04 0.39555555555555555 8 2019-02-07 0.38125000000000003 8 2018-12-31 0.17875000000000002 8 2018-12-27 0.25125 7 2019-01-25 -0.08857142857142856 7 2019-01-10 -0.0671428571428571 7 2019-02-02 -0.3942857142857143 7 2019-03-15 0.33714285714285713 7 2019-01-17 0.16285714285714287 7 2019-02-25 0.3228571428571429 7 2018-12-09 0.2414285714285714 7 2019-01-11 -0.25142857142857145 7 2019-01-14 0.1457142857142857 6 2019-02-26 0.08666666666666667 6 2019-04-05 0.25 6 2019-02-05 0.325 6 2018-12-24 0.11333333333333333 6 2018-12-13 0.03833333333333333 6 2019-01-12 0.18333333333333332 6 2018-12-17 0.42333333333333334 6 2019-04-02 0.26500000000000007 5 2019-03-11 -0.06200000000000001 5 2019-01-15 0.18999999999999997 5 2019-01-01 0.22200000000000003 5 2019-01-02 -0.21000000000000002 5 2019-03-08 0.41 5 2019-03-23 0.046 5 2018-12-15 0.146 5 2019-01-07 -0.40199999999999997 5 2019-03-24 0.35 5 2018-11-23 0.28400000000000003 4 2019-03-16 0.295 4 2019-02-28 0.13499999999999998 4 2019-02-20 -0.06 4 2018-11-13 -0.175 4 2019-02-18 0.4575 4 2018-12-14 -0.45999999999999996 4 2019-03-01 0.135 4 2019-02-03 0.4 3 2019-01-22 -0.27999999999999997 3 2019-01-06 0.07666666666666667 3 2019-02-11 0.26333333333333336 3 2019-01-20 -0.19333333333333333 3 2018-12-02 0.16666666666666666 3 2019-03-05 0.07333333333333335 3 2019-02-19 0.13333333333333333 3 2018-11-15 -0.12 3 2018-11-12 -0.17333333333333334 3 2018-11-14 -0.23666666666666666 3 2019-01-16 0.4566666666666667 3 2019-02-15 0.24333333333333332 3 2019-04-12 -0.45333333333333337 2 2019-03-09 0.4 2 2019-01-19 0.020000000000000018 2 2018-12-28 0.5599999999999999 2 2019-02-21 0.325 2 2019-02-14 -0.025000000000000022 2 2019-02-12 -0.065 2 2019-03-31 0.495 2 2018-12-16 0.545 2 2018-12-22 0.655 2 2019-04-07 0.6799999999999999 2 2019-03-04 0.29000000

In [23]:
%sql 

select ticker ,title, date(publishTime), sentiment from ngcdemo.appl_sentiments where date(publishTime) = '2019-03-25'

ticker title publishTime sentiment AAPL Jennifer Aniston Reese Witherspoon and Steve Carell preview The Morning Show at Apple launch event 2019-03-25 0.34 AAPL Why Roku Was Also a Big Winner After Apple?s Special TV Launch Event 2019-03-25 0.76 AAPL Apple Enters the Credit Card Market With?Yep?Apple Card 2019-03-25 0.38 AAPL How to Sign Up For an Apple News+ Subscription 2019-03-25 0.57 AAPL LAUGHABLE: Apple?s Big Idea Is? a Credit Card? 2019-03-25 0.8 AAPL Apple Releases macOS Mojave 10.14.4 With Safari Automatic Dark Mode and Apple News+ Support 2019-03-25 0.33 AAPL Why Today?s Apple Event Was So Unusual 2019-03-25 0.64 AAPL iOS 12.2 Suggests ECG App May Be Coming to UK and Other European Countries With watchOS 5.2 2019-03-25 0.44 AAPL Here are the top 6 announcements from Apple?s ?Show Time? event 2019-03-25 0.56 AAPL First pictures of Apple?s truly wireless Powerbeats Pro leak 2019-03-25 0.18 AAPL All the shows Apple wants you to get excited about 2019-03-25 0.64 AAPL It's time to accept that maybe Apple AirPower doesn't exist after all 2019-03-25 0.54 AAPL Apple unveils new streaming platform Apple TV+ 2019-03-25 0.53 AAPL The 4 Biggest Things Apple Just Announced From TV to a Credit Card 2019-03-25 0.86 AAPL S&P 500 slips with Apple lingering fears on global growth 2019-03-25 -0.64 AAPL All of Apple's new service announcements video - CNET 2019-03-25 0.61 AAPL All of Apple's new service announcements 2019-03-25 0.61 AAPL iOS 12.2 arrives with Apple News+ AirPods 2 support new Animoji 2019-03-25 0.86 AAPL Apple's big services announcement was strangely lacking in critical details 2019-03-25 0.1 AAPL Oprah Winfrey announces Apple TV+ documentaries and 'stimulating' book club on new streaming service 2019-03-25 0.69 AAPL US stocks end flat amid lingering growth doubts 2019-03-25 0.36 AAPL Apple TV+ joins crowded market for streaming options 2019-03-25 -0.05 AAPL Hollywood elite line up to support Apple TV+ in sleek originals promo 2019-03-25 0.72 AAPL Here?s everything Apple announced today 2019-03-25 0.0 AAPL Apple TV Plus with shows from Spielberg Oprah and J.J. Abrams is coming this fall - CNET 2019-03-25 0.0 AAPL This is Apple News+ the new app that includes magazine and newspaper subscriptions for $10 a month 2019-03-25 0.0 AAPL Sesame Street AppleTV+ Series ?Helpsters? Announced By Big Bird & Cody At Tech Giant Confab 2019-03-25 0.0 AAPL Oprah joins the Apple family with two new documentaries 2019-03-25 0.0 AAPL Apple News+ TV+ and more: How much do the new services cost and when are they available? 2019-03-25 0.0 AAPL Apple TV+ Announcement Raised More Questions Than It Answered 2019-03-25 0.18 AAPL Apple TV Plus is Apple?s Netflix challenger ? here?s what we know so far 2019-03-25 0.13 AAPL Oprah Winfrey Announces Two Apple Docs New Book Club! 2019-03-25 0.0 AAPL All the games confirmed for Apple Arcade a new subscription service coming this fall 2019-03-25 0.36 AAPL Here's the Full Rundown on Apple TV and the new Apple TV+ 2019-03-25 0.74 AAPL Apple Shares All of the Videos From Today's Services Event Introducing Apple News+ Apple TV+ Apple Arcade and Apple Card 2019-03-25 0.54 AAPL Here Are All the TV Shows Coming to Apple?s Streaming Service 2019-03-25 0.42 AAPL iOS 12.2 is out now with Apple News+ support 2019-03-25 -0.12 AAPL Why Oprah blessed Apple with her presence 2019-03-25 0.49 AAPL Apple?s New TV and Movie Streaming Service: What You Need to Know 2019-03-25 0.32 AAPL Apple TV+ video streaming service Apple News+ and Apple Arcade are official 2019-03-25 0.13 AAPL Reese Witherspoon Jennifer Aniston Reveal New Details About The Morning Show on Apple TV+ 2019-03-25 0.44 AAPL Apple just announced a slew of original TV shows ? check them out (AAPL) 2019-03-25 0.32 AAPL Watch Apple debut its game-subscription service Apple Arcade 2019-03-25 0.3 AAPL Oprah will release two documentaries on Apple TV Plus to ?heal our divisions? 2019-03-25 -0.34 AAPL Apple TV+ video streaming service to launch this fall 2019-03-25 

In [24]:
%sql 

select    Sticker ,Open, High, Low, Close, Volume, temp.senti as avg_sentiment  FROM generic_stocks , 
(select count(title) as cnt, date(publishTime) as pdate, avg(sentiment) as senti from ngcdemo.appl_sentiments group by date(publishTime) order by count(title) desc limit 1) temp 
where date(Date) = temp.pdate and Sticker = 'APPL';


Sticker Open High Low Close Volume avg_sentiment APPL 191.51 191.976 186.6 188.74 43326124 0.35960784313725486

## DELTA

Create a table called `stocks_delta` using `DELTA` out of the above data.

In [26]:

spark.sql("""
  USE ngcdemo
""")

spark.sql("""
  DROP TABLE IF EXISTS stocks_delta
""")
spark.sql("""
  CREATE TABLE stocks_delta
  USING DELTA 
  LOCATION '{}' 
""".format(deltaDataPath))
None

#### Metadata

Since we already have data backing `stocks_delta` in place, 
the table in the Hive metastore automatically inherits the schema, partitioning, 
and table properties of the existing data. 

Note that we only store table name, path, database info in Hive metastore,
the actual schema is stored in `_delta_logs`.

Metadata is displayed through `DESCRIBE DETAIL <tableName>`.

As long as we have some data in place already for a Delta table, we can infer schema.

In [28]:
%sql
DESCRIBE  stocks_delta

col_name data_type comment date timestamp null open double null high double null low double null close double null volume int null month int null Year int null sticker string null name string null # Partition Information # col_name data_type comment Year int null

-sandbox
Perform a simple `count` query to verify the number of records.

<img alt="Caution" title="Caution" style="vertical-align: text-bottom; position: relative; height:1.3em; top:0.0em" src="https://files.training.databricks.com/static/images/icon-warning.svg"/> Notice how the count is right off the bat; no need to worry about table repairs.

In [30]:
%sql
SELECT count(*) FROM generic_stocks

count(1) 5503

## Append

Now - Let's append some NetFlix Stock Data

In [32]:
janDataDF = (spark       
  .read                                              # Read a DataFrame from storage
  .option("inferSchema","true")                      # Infer schema
  .option("header","true")                           # File has a header
  .csv("dbfs:/databricks/ganeshrj/apple/NFLX_JAN.csv")     # Path to file 
  .drop("Adj Close"))

janDataDF = janDataDF.withColumn("Month", month(janDataDF.Date)).withColumn("Year",(year(janDataDF.Date))).withColumn("Stickers", lit("NFLX")) 

display(janDataDF)

Date Open High Low Close Volume Month Year Stickers 2018-01-02T00:00:00.000+0000 170.160004 172.300003 169.259995 172.259995 25555900 1 2018 NFLX 2018-01-03T00:00:00.000+0000 172.529999 174.550003 171.960007 172.229996 29517900 1 2018 NFLX 2018-01-04T00:00:00.000+0000 172.539993 173.470001 172.080002 173.029999 22434600 1 2018 NFLX 2018-01-05T00:00:00.000+0000 173.440002 175.369995 173.050003 175.0 23660000 1 2018 NFLX 2018-01-08T00:00:00.000+0000 174.350006 175.610001 173.929993 174.350006 20567800 1 2018 NFLX 2018-01-09T00:00:00.000+0000 174.550003 175.059998 173.410004 174.330002 21584000 1 2018 NFLX 2018-01-10T00:00:00.000+0000 173.160004 174.300003 173.0 174.289993 23959900 1 2018 NFLX 2018-01-11T00:00:00.000+0000 174.589996 175.490005 174.490005 175.279999 18667700 1 2018 NFLX 2018-01-12T00:00:00.000+0000 176.179993 177.360001 175.649994 177.089996 25418100 1 2018 NFLX 2018-01-16T00:00:00.000+0000 177.899994 179.389999 176.139999 176.190002 29565900 1 2018 NFLX 2018-01-17T00:00:00.000+0000 176.149994 179.25 175.070007 179.100006 33888500 1 2018 NFLX 2018-01-18T00:00:00.000+0000 179.369995 180.100006 178.25 179.259995 31193400 1 2018 NFLX 2018-01-19T00:00:00.000+0000 178.610001 179.580002 177.410004 178.460007 32425100 1 2018 NFLX 2018-01-22T00:00:00.000+0000 177.300003 177.779999 176.600006 177.0 27108600 1 2018 NFLX 2018-01-23T00:00:00.000+0000 177.300003 179.440002 176.820007 177.039993 32689100 1 2018 NFLX 2018-01-24T00:00:00.000+0000 177.25 177.300003 173.199997 174.220001 51105100 1 2018 NFLX 2018-01-25T00:00:00.000+0000 174.509995 174.949997 170.529999 171.110001 41529000 1 2018 NFLX 2018-01-26T00:00:00.000+0000 172.0 172.0 170.059998 171.509995 39143000 1 2018 NFLX 2018-01-29T00:00:00.000+0000 170.160004 170.160004 167.070007 167.960007 50640400 1 2018 NFLX 2018-01-30T00:00:00.000+0000 165.529999 167.369995 164.699997 166.970001 46048200 1 2018 NFLX 2018-01-31T00:00:00.000+0000 166.869995 168.440002 166.5 167.429993 32478900 1 2018 NFLX

Do a simple count of number of new items to be added to production data.

In [34]:
janDataDF.count()

Out[ 86 ]: 21

## APPEND Using Non-Delta pipeline
Append to the production table.

In the next cell, load the new data in `parquet` format and save to `../generic/aapl-data/`.

In [36]:
(janDataDF
  .write
  .format("parquet")
  .partitionBy("Year")
  .mode("append")
  .save(genericDataPath)
)

Query should show new results.

Should be `5503 + 21 = 5524` rows

In [38]:
%sql

select count(*) from generic_stocks

count(1) 5503

-sandbox

That's the not the right count (the new rows haven't been added in correctly). 

<img alt="Caution" title="Caution" style="vertical-align: text-bottom; position: relative; height:1.3em; top:0.0em" src="https://files.training.databricks.com/static/images/icon-warning.svg"/> Repair the table again and count the number of records.

In [40]:
%sql
MSCK REPAIR TABLE generic_stocks;

SELECT count(*) FROM generic_stocks

count(1) 5524

## APPEND Using Delta Pipeline

Next, repeat the process by writing to Delta format. 

In the next cell, load the new data in Delta format and save to `/delta/aapl-data/`.

In [42]:
%sql 

describe stocks_delta;

col_name data_type comment date timestamp null open double null high double null low double null close double null volume int null month int null Year int null sticker string null name string null # Partition Information # col_name data_type comment Year int null

In [43]:
(janDataDF
  .write
  .format("delta")
  .partitionBy("Year")
  .mode("append")
  .save(deltaDataPath)
)

In [44]:
janDataDF = (spark       
  .read                                              # Read a DataFrame from storage
  .option("inferSchema","true")                      # Infer schema
  .option("header","true")                           # File has a header
  .csv("dbfs:/databricks/ganeshrj/apple/NFLX_JAN.csv")     # Path to file 
  .drop("Adj Close"))

janDataDF = janDataDF.withColumn("Month", month(janDataDF.Date)).withColumn("Year",(year(janDataDF.Date))).withColumn("sticker1", lit("NFLX")) 

## Join with StocksDF to get the company name
janDataDF=janDataDF.join(broadcast(stocksDF), janDataDF.sticker1 == stocksDF.sticker).drop("sticker1") 

display(janDataDF)

Date Open High Low Close Volume Month Year sticker name 2018-01-02T00:00:00.000+0000 170.160004 172.300003 169.259995 172.259995 25555900 1 2018 NFLX Netflix Inc 2018-01-03T00:00:00.000+0000 172.529999 174.550003 171.960007 172.229996 29517900 1 2018 NFLX Netflix Inc 2018-01-04T00:00:00.000+0000 172.539993 173.470001 172.080002 173.029999 22434600 1 2018 NFLX Netflix Inc 2018-01-05T00:00:00.000+0000 173.440002 175.369995 173.050003 175.0 23660000 1 2018 NFLX Netflix Inc 2018-01-08T00:00:00.000+0000 174.350006 175.610001 173.929993 174.350006 20567800 1 2018 NFLX Netflix Inc 2018-01-09T00:00:00.000+0000 174.550003 175.059998 173.410004 174.330002 21584000 1 2018 NFLX Netflix Inc 2018-01-10T00:00:00.000+0000 173.160004 174.300003 173.0 174.289993 23959900 1 2018 NFLX Netflix Inc 2018-01-11T00:00:00.000+0000 174.589996 175.490005 174.490005 175.279999 18667700 1 2018 NFLX Netflix Inc 2018-01-12T00:00:00.000+0000 176.179993 177.360001 175.649994 177.089996 25418100 1 2018 NFLX Netflix Inc 2018-01-16T00:00:00.000+0000 177.899994 179.389999 176.139999 176.190002 29565900 1 2018 NFLX Netflix Inc 2018-01-17T00:00:00.000+0000 176.149994 179.25 175.070007 179.100006 33888500 1 2018 NFLX Netflix Inc 2018-01-18T00:00:00.000+0000 179.369995 180.100006 178.25 179.259995 31193400 1 2018 NFLX Netflix Inc 2018-01-19T00:00:00.000+0000 178.610001 179.580002 177.410004 178.460007 32425100 1 2018 NFLX Netflix Inc 2018-01-22T00:00:00.000+0000 177.300003 177.779999 176.600006 177.0 27108600 1 2018 NFLX Netflix Inc 2018-01-23T00:00:00.000+0000 177.300003 179.440002 176.820007 177.039993 32689100 1 2018 NFLX Netflix Inc 2018-01-24T00:00:00.000+0000 177.25 177.300003 173.199997 174.220001 51105100 1 2018 NFLX Netflix Inc 2018-01-25T00:00:00.000+0000 174.509995 174.949997 170.529999 171.110001 41529000 1 2018 NFLX Netflix Inc 2018-01-26T00:00:00.000+0000 172.0 172.0 170.059998 171.509995 39143000 1 2018 NFLX Netflix Inc 2018-01-29T00:00:00.000+0000 170.160004 170.160004 167.070007 167.960007 50640400 1 2018 NFLX Netflix Inc 2018-01-30T00:00:00.000+0000 165.529999 167.369995 164.699997 166.970001 46048200 1 2018 NFLX Netflix Inc 2018-01-31T00:00:00.000+0000 166.869995 168.440002 166.5 167.429993 32478900 1 2018 NFLX Netflix Inc

In [45]:
%sql
SELECT count(*) FROM stocks_delta

count(1) 5524

## UPSERT

## UPSERT Using Non-Delta Pipeline 

This feature is not supported in non-Delta pipelines.

Furthermore, no updates or deletes, or any sort of modifications are supported in non-Delta pipelines.

## UPSERT Using Delta Pipeline

## Part 4: Update and Insert Data

**Databricks Delta** delivers a powerful transactional storage layer by harnessing the power of Apache Spark and Databricks DBFS. The core abstraction of Databricks Delta is an optimized Spark table that stores data as Parquet files in DBFS.

Maintains a transaction log that efficiently tracks changes to the table.

You read and write data stored in the delta format using the same familiar Apache Spark SQL batch and streaming APIs that you use to work with Hive tables and DBFS directories. With the addition of the transaction log and other enhancements, Databricks Delta offers significant benefits:

#### ACID transactions
Multiple writers can simultaneously modify a dataset and see consistent views. For qualifications, see Multi-cluster writes.
Writers can modify a dataset without interfering with jobs reading the dataset.
#### Fast read access
Automatic file management organizes data into large files that can be read efficiently.
Statistics enable speeding up reads by 10-100x and and data skipping avoids reading irrelevant information.

#### Upserts (MERGE INTO)
The MERGE INTO statement allows you to merge a set of updates and insertions into an existing dataset.

In [50]:
dbutils.fs.rm(deltaMiniDataPath, True)
deltaMiniDataPath = userhome + "/delta/mini-delta1"
newJanDataDF = janDataDF.where((col("Date") < '2018-01-13T00:00:00.000+0000')   &  (col("sticker") == 'NFLX'))

(newJanDataDF
  .write
  .format("delta")
  .partitionBy("Year")
  .save(deltaMiniDataPath) 
)

display (newJanDataDF)
spark.sql("use ngcdemo")
spark.sql("""
    DROP TABLE IF EXISTS jan_nflx_delta_mini
  """)
spark.sql("""
    CREATE TABLE jan_nflx_delta_mini
    USING DELTA 
    LOCATION '{}' 
  """.format(deltaMiniDataPath))
None

Date Open High Low Close Volume Month Year sticker name 2018-01-02T00:00:00.000+0000 170.160004 172.300003 169.259995 172.259995 25555900 1 2018 NFLX Netflix Inc 2018-01-03T00:00:00.000+0000 172.529999 174.550003 171.960007 172.229996 29517900 1 2018 NFLX Netflix Inc 2018-01-04T00:00:00.000+0000 172.539993 173.470001 172.080002 173.029999 22434600 1 2018 NFLX Netflix Inc 2018-01-05T00:00:00.000+0000 173.440002 175.369995 173.050003 175.0 23660000 1 2018 NFLX Netflix Inc 2018-01-08T00:00:00.000+0000 174.350006 175.610001 173.929993 174.350006 20567800 1 2018 NFLX Netflix Inc 2018-01-09T00:00:00.000+0000 174.550003 175.059998 173.410004 174.330002 21584000 1 2018 NFLX Netflix Inc 2018-01-10T00:00:00.000+0000 173.160004 174.300003 173.0 174.289993 23959900 1 2018 NFLX Netflix Inc 2018-01-11T00:00:00.000+0000 174.589996 175.490005 174.490005 175.279999 18667700 1 2018 NFLX Netflix Inc 2018-01-12T00:00:00.000+0000 176.179993 177.360001 175.649994 177.089996 25418100 1 2018 NFLX Netflix Inc

## Fix the data issue

In [52]:
from pyspark.sql.functions import lit
correctedDataDF = (newJanDataDF
  .withColumn("Volume", lit(4000000))
 )

spark.sql("use ngcdemo")
spark.sql("DROP TABLE IF EXISTS corrected_nflx_delta_to_upsert")
correctedDataDF.write.saveAsTable("corrected_nflx_delta_to_upsert")

In [53]:
%sql select * from corrected_nflx_delta_to_upsert

Date Open High Low Close Volume Month Year sticker name 2018-01-02T00:00:00.000+0000 170.160004 172.300003 169.259995 172.259995 4000000 1 2018 NFLX Netflix Inc 2018-01-03T00:00:00.000+0000 172.529999 174.550003 171.960007 172.229996 4000000 1 2018 NFLX Netflix Inc 2018-01-04T00:00:00.000+0000 172.539993 173.470001 172.080002 173.029999 4000000 1 2018 NFLX Netflix Inc 2018-01-05T00:00:00.000+0000 173.440002 175.369995 173.050003 175.0 4000000 1 2018 NFLX Netflix Inc 2018-01-08T00:00:00.000+0000 174.350006 175.610001 173.929993 174.350006 4000000 1 2018 NFLX Netflix Inc 2018-01-09T00:00:00.000+0000 174.550003 175.059998 173.410004 174.330002 4000000 1 2018 NFLX Netflix Inc 2018-01-10T00:00:00.000+0000 173.160004 174.300003 173.0 174.289993 4000000 1 2018 NFLX Netflix Inc 2018-01-11T00:00:00.000+0000 174.589996 175.490005 174.490005 175.279999 4000000 1 2018 NFLX Netflix Inc 2018-01-12T00:00:00.000+0000 176.179993 177.360001 175.649994 177.089996 4000000 1 2018 NFLX Netflix Inc

In [54]:
%sql
MERGE INTO stocks_delta
USING corrected_nflx_delta_to_upsert
ON corrected_nflx_delta_to_upsert.Date = stocks_delta.Date
WHEN MATCHED THEN
  UPDATE SET
    stocks_delta.Volume = corrected_nflx_delta_to_upsert.Volume
WHEN NOT MATCHED
  THEN INSERT (Date, Open, High, Low, Close, Volume, Month, Year, sticker, name)
  VALUES (
    corrected_nflx_delta_to_upsert.Date,
    corrected_nflx_delta_to_upsert.Open,
    corrected_nflx_delta_to_upsert.High,
    corrected_nflx_delta_to_upsert.Low,
    corrected_nflx_delta_to_upsert.Close,
    corrected_nflx_delta_to_upsert.Volume,
    corrected_nflx_delta_to_upsert.Month,
    corrected_nflx_delta_to_upsert.Year,
    corrected_nflx_delta_to_upsert.sticker,
    corrected_nflx_delta_to_upsert.name)

In [55]:
%sql
SELECT * FROM stocks_delta
where Date < '2018-01-13T00:00:00.000+0000' and sticker = 'NFLX'

date open high low close volume month Year sticker name 2018-01-08T00:00:00.000+0000 174.350006 175.610001 173.929993 174.350006 4000000 1 2018 NFLX null 2018-01-09T00:00:00.000+0000 174.550003 175.059998 173.410004 174.330002 4000000 1 2018 NFLX null 2018-01-04T00:00:00.000+0000 172.539993 173.470001 172.080002 173.029999 4000000 1 2018 NFLX null 2018-01-05T00:00:00.000+0000 173.440002 175.369995 173.050003 175.0 4000000 1 2018 NFLX null 2018-01-10T00:00:00.000+0000 173.160004 174.300003 173.0 174.289993 4000000 1 2018 NFLX null 2018-01-03T00:00:00.000+0000 172.529999 174.550003 171.960007 172.229996 4000000 1 2018 NFLX null 2018-01-02T00:00:00.000+0000 170.160004 172.300003 169.259995 172.259995 4000000 1 2018 NFLX null 2018-01-11T00:00:00.000+0000 174.589996 175.490005 174.490005 175.279999 4000000 1 2018 NFLX null 2018-01-12T00:00:00.000+0000 176.179993 177.360001 175.649994 177.089996 4000000 1 2018 NFLX null

## Timetravel

In [57]:
%sql 

describe history stocks_delta;

version timestamp userId userName operation operationParameters job notebook clusterId readVersion isolationLevel 5 2019-04-29T02:24:22.000+0000 100959 ganesh.rajagopal@databricks.com MERGE Map(predicate -> (ngcdemo.corrected_nflx_delta_to_upsert.`Date` = ngcdemo.stocks_delta.`Date`)) null List(2911691) 0426-204644-shear440 4 WriteSerializable 4 2019-04-29T01:44:46.000+0000 100959 ganesh.rajagopal@databricks.com WRITE Map(mode -> Append, partitionBy -> ["Year"]) null List(2911691) 0426-204644-shear440 3 WriteSerializable 3 2019-04-29T00:38:54.000+0000 100959 ganesh.rajagopal@databricks.com WRITE Map(mode -> Append, partitionBy -> ["Year"]) null List(2911691) 0426-204644-shear440 2 WriteSerializable 2 2019-04-29T00:38:47.000+0000 100959 ganesh.rajagopal@databricks.com WRITE Map(mode -> Append, partitionBy -> ["Year"]) null List(2911691) 0426-204644-shear440 1 WriteSerializable 1 2019-04-29T00:38:32.000+0000 100959 ganesh.rajagopal@databricks.com WRITE Map(mode -> Append, partitionBy -> ["Year"]) null List(2911691) 0426-204644-shear440 0 WriteSerializable 0 2019-04-29T00:38:17.000+0000 100959 ganesh.rajagopal@databricks.com WRITE Map(mode -> Append, partitionBy -> ["Year"]) null List(2911691) 0426-204644-shear440 null WriteSerializable

In [58]:
%sql 

SELECT count(*) FROM stocks_delta VERSION AS OF 1 where volume = 4000000;

count(1) 0

In [59]:
%sql 

SELECT count(*) FROM stocks_delta VERSION AS OF 5 where volume = 4000000;

count(1) 36

In [60]:
%sql 

select count(*) from stocks_delta   TIMESTAMP AS OF "2019-04-12 13:53:01"    where volume = 4000000;

com.databricks.backend.common.rpc.DatabricksExceptions$SQLExecutionException: org.apache.spark.sql.AnalysisException: Table or view not found: aapl_delta; line 1 pos 21
	at org.apache.spark.sql.catalyst.analysis.package$AnalysisErrorAt.failAnalysis(package.scala:47)
	at org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveRelations$.org$apache$spark$sql$catalyst$analysis$Analyzer$ResolveRelations$$lookupTableFromCatalog(Analyzer.scala:750)
	at org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveRelations$.resolveRelation(Analyzer.scala:695)
	at org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveRelations$$anonfun$apply$8.applyOrElse(Analyzer.scala:731)
	at org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveRelations$$anonfun$apply$8.applyOrElse(Analyzer.scala:724)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$$anonfun$resolveOperatorsUp$1$$anonfun$apply$1.apply(AnalysisHelper.scala:90)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$$anonfun$resolveOperatorsUp$1$$anonfun$apply$1.apply(AnalysisHelper.scala:90)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:77)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$$anonfun$resolveOperatorsUp$1.apply(AnalysisHelper.scala:89)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$$anonfun$resolveOperatorsUp$1.apply(AnalysisHelper.scala:86)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$.allowInvokingTransformsInAnalyzer(AnalysisHelper.scala:194)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$class.resolveOperatorsUp(AnalysisHelper.scala:86)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.resolveOperatorsUp(LogicalPlan.scala:29)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$$anonfun$resolveOperatorsUp$1$$anonfun$1.apply(AnalysisHelper.scala:87)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$$anonfun$resolveOperatorsUp$1$$anonfun$1.apply(AnalysisHelper.scala:87)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$8.apply(TreeNode.scala:351)
	at org.apache.spark.sql.catalyst.trees.TreeNode.mapProductIterator(TreeNode.scala:208)
	at org.apache.spark.sql.catalyst.trees.TreeNode.mapChildren(TreeNode.scala:349)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$$anonfun$resolveOperatorsUp$1.apply(AnalysisHelper.scala:87)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$$anonfun$resolveOperatorsUp$1.apply(AnalysisHelper.scala:86)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$.allowInvokingTransformsInAnalyzer(AnalysisHelper.scala:194)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$class.resolveOperatorsUp(AnalysisHelper.scala:86)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.resolveOperatorsUp(LogicalPlan.scala:29)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$$anonfun$resolveOperatorsUp$1$$anonfun$1.apply(AnalysisHelper.scala:87)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$$anonfun$resolveOperatorsUp$1$$anonfun$1.apply(AnalysisHelper.scala:87)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$8.apply(TreeNode.scala:351)
	at org.apache.spark.sql.catalyst.trees.TreeNode.mapProductIterator(TreeNode.scala:208)
	at org.apache.spark.sql.catalyst.trees.TreeNode.mapChildren(TreeNode.scala:349)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$$anonfun$resolveOperatorsUp$1.apply(AnalysisHelper.scala:87)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$$anonfun$resolveOperatorsUp$1.apply(AnalysisHelper.scala:86)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$.allowInvokingTransformsInAnalyzer(AnalysisHelper.scala:194)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$class.resolveOperatorsUp(AnalysisHelper.scala:86)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.resolveOperatorsUp(LogicalPlan.scala:29)
	at org.apache.spark.sql.catalyst.pla

In [61]:
%sql 

Optimize stocks_delta  
   zorder by Date

path null

In [62]:
%sql
SELECT * FROM stocks_delta
where Date < '2018-01-13T00:00:00.000+0000' and sticker = 'NFLX'

date open high low close volume month Year sticker name 2018-01-12T00:00:00.000+0000 176.179993 177.360001 175.649994 177.089996 4000000 1 2018 NFLX null 2018-01-02T00:00:00.000+0000 170.160004 172.300003 169.259995 172.259995 4000000 1 2018 NFLX null 2018-01-11T00:00:00.000+0000 174.589996 175.490005 174.490005 175.279999 4000000 1 2018 NFLX null 2018-01-08T00:00:00.000+0000 174.350006 175.610001 173.929993 174.350006 4000000 1 2018 NFLX null 2018-01-09T00:00:00.000+0000 174.550003 175.059998 173.410004 174.330002 4000000 1 2018 NFLX null 2018-01-04T00:00:00.000+0000 172.539993 173.470001 172.080002 173.029999 4000000 1 2018 NFLX null 2018-01-05T00:00:00.000+0000 173.440002 175.369995 173.050003 175.0 4000000 1 2018 NFLX null 2018-01-03T00:00:00.000+0000 172.529999 174.550003 171.960007 172.229996 4000000 1 2018 NFLX null 2018-01-10T00:00:00.000+0000 173.160004 174.300003 173.0 174.289993 4000000 1 2018 NFLX null

## Part 5: Streaming Data - combine live and historical data (without the complexity of traditional lambda architecture)

In Databricks, incorporating real-time or near-real-time data is just as simple as incorporating static data. Let's look at how we can combine a streaming source, in this case eProcessing Events, with our existing data, using the power of Delta and Databricks Runtime.

<div style="text-align: center; line-height: 0; padding-top: 9px;"><br><img src="https://i.imgur.com/RYOs8lB.png" height="00" width="700"></div>

In [64]:
streamingDataPath = userhome + "/delta/streaming/"
eventStreamPath = userhome + "/delta/eventStream"
dbutils.fs.rm(streamingDataPath, True)

Out[ 13 ]: True

In [65]:
from datetime import datetime
def createEvents( count, date, eventType, open1, high, low, close, month, year, sticker, name ):
  return [(datetime.strptime(date, '%Y-%m-%d'), eventType, open1, high, low, close, month, year, sticker, name , ((i%5)+1) * count ) for i in range(count)]

In [66]:
from pyspark.sql.types import *
schemaFields = [StructField('date', DateType(), True), 
                StructField('eventType', StringType(), True), 
                StructField('open1', DoubleType(), True), 
                StructField('high', DoubleType(), True), 
                StructField('low', DoubleType(), True), 
                StructField('close', DoubleType(), True),
                StructField('month', IntegerType(), True), 
                StructField('year', IntegerType(), True),
                StructField('sticker', StringType(), True), 
                StructField('name', StringType(), True),               
                StructField('volume', IntegerType(), True)]
schema = StructType(schemaFields)

## Simulate a Batch Insert

In [68]:
insertDataDF = spark.createDataFrame(createEvents( 1000, "2017-10-01", "batch-insert", 123.33, 128.33, 121.40, 124.33, 10, 2017, 'NFLX', 'Netflix Inc'), schema)
insertDataDF.write.format("delta").save(streamingDataPath)


In [69]:
display (insertDataDF)

date eventType open1 high low close month year sticker name volume 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 1000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 2000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 3000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 4000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 5000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 1000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 2000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 3000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 4000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 5000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 1000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 2000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 3000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 4000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 5000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 1000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 2000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 3000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 4000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 5000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 1000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 2000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 3000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 4000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 5000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 1000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 2000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 3000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 4000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 5000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 1000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 2000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 3000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 4000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 5000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 1000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 2000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 3000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 4000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 5000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 1000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 2000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 3000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 4000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 5000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 1000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 2000 2017-10-01 batch-insert 123.33 128.33 121.4 124.33 10 2017 NFLX Netflix Inc 3000 2017-10-01 batch-insert 123.33 128.33 121.4 1

## Simulate a Batch Append for couple of Stocks

In [71]:
(spark.createDataFrame(createEvents(5000, "2017-10-02", "batch-append-1", 123.33, 128.33, 121.40, 124.33, 10, 2017, 'APPL', 'Apple Computers Inc'), schema)
  .write
  .format("delta")
  .mode("append")
  .save(str(streamingDataPath)))

In [72]:
(spark.createDataFrame(createEvents(5000, "2017-10-03", "batch-append-2", 123.33, 128.33, 121.40, 124.33, 10, 2017, 'GOOG', 'Google Inc'), schema)
  .write
  .format("delta")
  .mode("append")
  .save(str(streamingDataPath)))

## Display the results

In [74]:
display(
  (spark
  .readStream
  .format("delta")
  .load(str(streamingDataPath))
  .groupBy("date","eventType", "sticker")
  .count()
  .orderBy("date"))
)

date eventType sticker count 2017-10-01 batch-insert NFLX 1000 2017-10-02 batch-append-1 APPL 5000 2017-10-03 batch-append-2 GOOG 5000 2017-10-10 stream NFLX 10000

## Kick the streaming job and watch the data being processed

In [76]:
df = spark.createDataFrame(createEvents(10000, "2017-10-10", "stream", 123.33, 128.33, 121.40, 124.33, 10, 2017, 'NFLX', 'Netflix Inc'), schema)
schemaNew = df.schema
(df.write
  .format("json")
  .mode("overwrite")
  .save(str(eventStreamPath)))

streamDF = spark.readStream.schema(schemaNew).option("maxFilesPerTrigger", 1).json(str(eventStreamPath))

(streamDF
  .writeStream
  .format("delta")
  .option("path", streamingDataPath)
  .option("checkpointLocation", "/tmp/checkpoint/")
  .start())

Out[ 20 ]: <pyspark.sql.streaming.StreamingQuery at 0x7f8fce5ab5c0>

## Summary:  

Databricks provides a transformative, best-in-class platform that unifies Data Science, Analytics, and Data Engineering to help solve the toughest business problems in the world.

In a few lines of code, we created a cross-silo 360 view of the user that included live streaming data; this data can be directly fed to Data Scientists, Analysts, or other users downstream who can immediately access and further manipulate it.

We just demonstrated that Databricks Delta supports batch reports, exploratory analytics, DW queries (such as aggregations on a  dataset),  and BI/interactive queries on a **single storage/copy of data**.